### ISSUE TITLE
Improve dynamic server count detect logic of agent

### ISSUE THREAD

**Issue by zqzten** [kubernetes-sigs/apiserver-network-proxy#358]
Currently if we want the agent to dynamic detect the server count change, we have to enable the `syncForever` mode, however, with this mode on, agent just call the `Connect` method of server constantly, and for most cases it just grap the sever count and close the connection (since the server count is not chaging that frequently). This cause a lot of garbage error log (specifically `server.go:767] "stream read failure" err="rpc error: code = Canceled desc = context canceled"` on server side) and a waste of networking and computing resource.

So would it be applicable that we introduce a lightweight method `ServerCount` to server and let agent call that method to sync server count instead of the heavier `Connect`?

**Comment by ipochi** [comment]
@cheftako @tallclair 

I'd be interested to know more about what others in the community feel regarding this and is there a scope/priority for a better solution ?

I'd like to collaborate with relevant folks and formulate an approach.

**Comment by tallclair** [comment]
I agree that the current approach is problematic. Introducing a new lightweight method makes sense to me, but I'd like someone with more experience on this project to weigh in.

**Comment by zqzten** [comment]
> One thing we haven't address yet is how the server gets that value.

Good point. I've noticed that there was another rotting issue #273 tracking on this. We may consider that together with this issue.

**Comment by jkh52** [comment]
> Good point. I've noticed that there was another rotting issue https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/273 tracking on this. We may consider that together with this issue.

We could keep this issue open, to improve "how the agents learn server count". There are some ideas here that weren't considered there. https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/273 can track the remaining server half of the feature.

**Comment by carreter** [comment]
FYI, if an agent receives an increased server count from a server it is already connected to, it doesn't reset the backoff and just uses the last set duration. I'm guessing that this isn't the desired behavior and that the agent should immediately reset the backoff and shift into "fast sync" mode if it is supposed to be connected to more servers than it currently is.

This will only be a problem once #273 is implemented as there is currently no way to update the server count on the server side.

**Comment by carreter** [comment]
Fixed by #643 !
